File: shared/src/main/java/org/pac4j/play/deadbolt2/Pac4jSubject.java
package org.pac4j.play.deadbolt2;

import be.objectify.deadbolt.java.models.Permission;
import be.objectify.deadbolt.java.models.Role;
import be.objectify.deadbolt.java.models.Subject;
import org.pac4j.core.profile.CommonProfile;

import java.util.ArrayList;
import java.util.List;

/**
 * The deadbolt subject built from the pac4j user profile.
 *
 * @author Jerome Leleu
 * @since 2.6.0
 */
public class Pac4jSubject implements Subject {

    private String id;

    private List<Pac4jRole> roles = new ArrayList<>();

    private List<Pac4jPermission> permissions = new ArrayList<>();

    public Pac4jSubject(final CommonProfile profile) {
        id = profile.getId();
        for (final String role : profile.getRoles()) {
            roles.add(new Pac4jRole(role));
        }
        for (final String permission : profile.getPermissions()) {
            permissions.add(new Pac4jPermission(permission));
        }
    }

    @Override
    public List<? extends Role> getRoles() {
        return roles;
    }

    @Override
    public List<? extends Permission> getPermissions() {
        return permissions;
    }

    @Override
    public String getIdentifier() {
        return id;
    }
}

File: shared/src/main/scala/org/pac4j/play/scala/deadbolt2/Pac4jSubject.scala
package org.pac4j.play.scala.deadbolt2

import scala.collection.JavaConverters._

import be.objectify.deadbolt.scala.models.{Permission, Role, Subject}
import org.pac4j.core.profile.CommonProfile

/**
  * @author Zenkie Zhu
  * @since 4.1.0
  */
case class Pac4jRole(
  name: String
) extends Role


case class Pac4jPermission(
  value: String
) extends Permission


case class Pac4jSubject(
  identifier: String,

  roles: List[Role],
  permissions: List[Permission],

  name: String,
  avatarURL: Option[String]
) extends Subject


object Pac4jSubject {
  def apply(profile: CommonProfile): Pac4jSubject = {
    Pac4jSubject(
      profile.getId,

      profile.getRoles.asScala.map(Pac4jRole(_)).toList,
      profile.getPermissions.asScala.map(Pac4jPermission(_)).toList,

      profile.getDisplayName,
      Option(profile.getPictureUrl).map(_.toString())
    )
  }
}
